src/ngraph/runtime/cpu/pass/cpu_rnn_mat_fusion.cpp

```diff
@@ -0,0 +1,288 @@
+/*******************************************************************************
+* Copyright 2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <fstream>
+#include <map>
+#include <memory>
+#include <numeric>
+#include <stack>
+#include <typeindex>
+#include <unordered_map>
+
+#include "cpu_rnn_mat_fusion.hpp"
+#include "ngraph/op/add.hpp"
+#include "ngraph/op/broadcast.hpp"
+#include "ngraph/op/dot.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/op/slice.hpp"
+
+using namespace ngraph;
+
+typedef std::shared_ptr<Node> NodePtr;
+
+#define TI(x) std::type_index(typeid(x))
+
+// a sequence of nodes, identified with a segment type for the input parameter type
+struct NodeSegment : public NodeVector
+{
+    enum Type
+    {
+        DATA = 0,
+        WEIGHTS,
+        BIAS,
+        UNDEFINED
+    };
+    Type type{UNDEFINED};
+};
+typedef std::pair<NodeSegment::Type, std::vector<std::type_index>> NodeTypeSequence;
+typedef std::list<NodeTypeSequence> NodeTypeSequenceList;
+
+// Preorder traversal to collect all valid segments in the graph
+// precondition: all valid sequences must be unique
+// [a, b, c] and [a, c, b] are different, for valid sequences like [a, b, c] and [a, b], the
+// longest sequence will be matched.
+void FindValidSegments(const NodePtr& node,
+                       NodeSegment segment,
```

Comment:
Heh, this is literally a matcher in reverse 😄 I'm kinda wondering if it might be easier to simply go over the entire graph and give a matcher your sequences (in the reverse order) if it returns a match -- excellent, you got a new sequence of a particular type and a node it starts from. if @jbobba doesn't mind we could spend some time to try to simplify this. As a bonus, the labels would map to parameters and we could be constructing some of the maps @louis constructs later on, in parallel w/ a graph walk.

Reply:
I'd replace a `preorder` traversal with `users' chain` or stn like dat.